odilo.py

```diff
@@ -0,0 +1,581 @@
+        library = DatabaseTest.make_default_library(_db)
+        collection, ignore = get_one_or_create(
+            _db, Collection,
+            name="Test Odilo Collection",
+            create_method_kwargs=dict(
+                external_account_id=u'library_id_123',
+            )
+        )
+        integration = collection.create_external_integration(
+            protocol=ExternalIntegration.ODILO
+        )
+        integration.password = u'abcdef123hijklm'
+        library.collections.append(collection)
+        return collection
+
+
+class OdiloRepresentationExtractor(object):
+    """Extract useful information from Odilo's JSON representations."""
+
+    log = logging.getLogger("OdiloRepresentationExtractor")
+
+    format_data_for_odilo_format = {
+        "ACSM": (
+            Representation.PDF_MEDIA_TYPE, DeliveryMechanism.ADOBE_DRM
+        ),
+        "PDF": (
+            Representation.PDF_MEDIA_TYPE, DeliveryMechanism.STREAMING_TEXT_CONTENT_TYPE
+        ),
+        "EPUB": (
+            Representation.EPUB_MEDIA_TYPE, DeliveryMechanism.STREAMING_TEXT_CONTENT_TYPE
+        ),
+        "MP3": (
+            Representation.MP3_MEDIA_TYPE, DeliveryMechanism.STREAMING_AUDIO_CONTENT_TYPE
+        ),
+        "MP4": (
+            Representation.MP4_MEDIA_TYPE, DeliveryMechanism.STREAMING_VIDEO_CONTENT_TYPE
+        ),
+        "WMV": (
+            Representation.WMV_MEDIA_TYPE, DeliveryMechanism.STREAMING_VIDEO_CONTENT_TYPE
+        ),
+        "JPG": (
+            Representation.JPEG_MEDIA_TYPE, DeliveryMechanism.NO_DRM
+        ),
+        "SCORM": (
+            Representation.ZIP_MEDIA_TYPE, DeliveryMechanism.NO_DRM
+        )
+    }
+
+    odilo_medium_to_simplified_medium = {
+        "ACSM": Edition.BOOK_MEDIUM,
+        "PDF": Edition.BOOK_MEDIUM,
+        "EPUB": Edition.BOOK_MEDIUM,
+        "MP3": Edition.AUDIO_MEDIUM,
+        "MP4": Edition.VIDEO_MEDIUM,
+        "WMV": Edition.VIDEO_MEDIUM,
+        "JPG": Edition.IMAGE_MEDIUM,
+        "SCORM": Edition.ELECTRONIC_FORMAT
+    }
+
+    @classmethod
+    def record_info_to_circulation(cls, availability):
+        """ Note:  The json data passed into this method is from a different file/stream 
+        from the json data that goes into the record_info_to_metadata() method.
+        """
+
+        if 'recordId' not in availability:
+            return None
+
+        record_id = availability['recordId']
+        primary_identifier = IdentifierData(Identifier.ODILO_ID, record_id)  # We own this availability.
+
+        licenses_owned = int(availability['totalCopies'])
+        licenses_available = int(availability['availableCopies'])
+        if 'numLoans' in availability:
+            licenses_checked_out = int(availability['numLoans'])
+        licenses_reserved = int(availability['holdsQueueSize'])
+        if 'numPatronsInHoldQueue' in availability:
+            patrons_in_hold_queue = int(availability['numPatronsInHoldQueue'])
```

Comment:
I want to be careful about this. `patrons_in_hold_queue` contains the number of patrons who are currently waiting for a copy of the book. `licenses_reserved` is the number of patrons who put the book on hold _earlier_, but who are now at the front of the queue and who could get the book right now if they wanted to. `licenses_reserved` should never be larger than `patrons_in_hold_queue`, because everyone with a reserved license is also in the hold queue.

It's not clear to me what `holdsQueueSize` is measuring or how it is different from `numPatronsInHoldQueue`, so I want to make sure you're mapping the information appropriately.

Reply:
ok, I have revised that, fixed it in the integration and updated the Connector availability API. Now we have:
- 'notifiedHolds' --> Patrons that placed a hold and have already been notified, so they can get the book when they want. 
- 'holdsQueueSize' --> all the patrons wainting for a copy of a specific book (notified or not).